packages/ui-app/src/InputNumber.tsx

```diff
@@ -0,0 +1,159 @@
+// Copyright 2017-2018 @polkadot/ui-app authors & contributors
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+import { TranslationFunction } from 'i18next';
+import { RawParam$OnChange } from './Params/types';
+import { BareProps } from './types';
+
+import BN from 'bn.js';
+import React from 'react';
+
+import isValidBalance from './util/isValidBalance';
+import { keydown } from './util/keyboard';
+import classes from './util/classes';
+import Input from './Input';
+import { KEYS_ALLOWED, KEYS_PRE } from './constants';
+
+type Props = BareProps & {
+  defaultValue?: string,
+  isError?: boolean,
+  label?: any,
+  maxLength?: number,
+  onChange?: RawParam$OnChange | undefined,
+  placeholder?: string,
+  t: TranslationFunction,
+  withLabel?: boolean
+};
+
+type State = {
+  error?: React.ReactNode,
+  info?: React.ReactNode,
+  warn?: React.ReactNode,
+  isPreKeyDown: boolean
+};
+
+class InputNumber extends React.PureComponent<Props, State> {
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+
+    this.state = {
+      error: '',
+      warn: '',
+      info: '',
+      isPreKeyDown: false
+    };
+  }
+
+  render () {
+    const { className, defaultValue, isError, label, maxLength, placeholder, style, withLabel } = this.props;
+    const { error, info, warn } = this.state;
+
+    return (
+      <div
+        className={classes('ui--InputNumber', className)}
+        style={style}
+      >
+        <Input
+          className={className}
+          defaultValue={defaultValue || '0'}
+          error={error}
+          info={info}
+          isError={isError}
+          label={label}
+          maxLength={maxLength}
+          onChange={this.onChange}
+          onKeyDown={this.onKeyDown}
+          onKeyUp={this.onKeyUp}
+          placeholder={placeholder}
+          style={style}
+          type='text'
+          warn={warn}
+          withLabel={withLabel}
+        />
+      </div>
+    );
+  }
+
+  onChange = (value: string): void => {
+    const { onChange, t } = this.props;
+
+    try {
+      const { isValid, errorMessage, infoMessage, warnMessage } = isValidBalance(value, t);
+
+      this.setState({
+        error: !isValid && errorMessage ? errorMessage : '',
+        info: isValid && infoMessage ? infoMessage : '',
+        warn: isValid && warnMessage ? warnMessage : ''
+      });
+
+      let valueBN: BN;
+      valueBN = new BN(value || '0');
+
+      if (!onChange) {
+        return;
+      }
+
+      onChange({
+        isValid,
```

Comment:
As per earlier comments on the actual onChange type - we should check the value, if valid, callback with _only_ the value, nothing more required.

Reply:
This is still outstanding. `onChange` should be - `onChange: (value: BN) => void`, not extra `inValid` is should be what it says on the tin, i.e. a number input.